mailpoet/lib/Subscription/Manage.php

```diff
@@ -132,34 +135,39 @@ public function onSave() {
     $this->urlHelper->redirectBack($result);
   }
 
-  private function updateSubscriptions(Subscriber $subscriber, array $subscriberData) {
+  private function updateSubscriptions(SubscriberEntity $subscriber, array $subscriberData): void {
     $segmentsIds = [];
     if (isset($subscriberData['segments']) && is_array($subscriberData['segments'])) {
       $segmentsIds = $subscriberData['segments'];
     }
-    $subscriber->withSubscriptions();
     $allowedSegments = $this->settings->get('subscription.segments', false);
 
     // Unsubscribe from all other segments already subscribed to
     // but don't change disallowed segments
-    foreach ($subscriber->subscriptions as $subscription) {
-      $segmentId = $subscription['segment_id'];
-      if ($allowedSegments && !in_array($segmentId, $allowedSegments)) {
+    foreach ($subscriber->getSubscriberSegments() as $subscriberSegment) {
+      $segment = $subscriberSegment->getSegment();
+      if (!$segment) {
+        $this->subscriberSegmentRepository->remove($subscriberSegment);
+        $subscriber->getSubscriberSegments()->removeElement($subscriberSegment);
```

Comment:
Is this necessary? I'm not very familiar with this code so I'm just trying to wrap my head around it. It seems like when we want to use the subscriber segments again we fetch them from the repository down on L168:171.

If the concern is about saving the subscriber with invalid segments it seems like [this code](https://github.com/mailpoet/mailpoet/blob/9867b8c7ad3e822836ebfca8df592c7a29debfc0/mailpoet/lib/Entities/SubscriberEntity.php#L524-L524) should take care of it.

I definitely don't mind having this in here if it's not strictly necessary! It's good to know that `$subscriber->getSubscriberSegments()` will be correct after this `foreach` loop. Again I just want to make sure I understand what's going on.

Reply:
Good point. I didn't realize that we have a code that solves it. I will remove it.